server/src/main/java/org/eclipse/lsp/cobol/service/CFASTBuilderImpl.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2021 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *    Broadcom, Inc. - initial API and implementation
+ *
+ */
+package org.eclipse.lsp.cobol.service;
+
+import org.eclipse.lsp.cobol.core.CobolParser;
+import org.eclipse.lsp.cobol.core.model.extendedapi.*;
+import org.eclipse.lsp.cobol.core.model.tree.*;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * CF tree builder iplementation
+ */
+public class CFASTBuilderImpl implements CFASTBuilder {
+  private static final int SNIPPET_LENGTH = 10;
+  //  IF statement without END_IF workaround. Should be removed after grammar fixed.
+  private boolean needCloseIf = false;
+
+  @Override
+  public ExtendedApiResult build(Node rootNode) {
+    ExtendedApiResult result = new ExtendedApiResult(new ArrayList<>());
+    if (rootNode == null) {
+      return result;
+    }
+    for (Node node : rootNode.getChildren()) {
+      traverse((ProgramNode) node, result.getControlFlowAST());
+    }
+    return result;
+  }
+
+  private void traverse(CFASTNode parent, Node node) {
+    if (node instanceof ParagraphNode) {
+      Paragraph paragraph = new Paragraph();
+      paragraph.setLocation(convertLocation(node.getLocality().toLocation()));
```

Comment:
Either extract as small method getParagraph(node)   or in Paragrapth define builder with chained methods so code will be shorter;